File: src/main/java/blynx/thymeleaf/compositiondialect/CompositionComponent.java
package blynx.thymeleaf.compositiondialect;

import java.util.Map;

/**
 * Every composition component implements this — a record whose components are its context and its props,
 * or a plain class implementing it directly (see {@link AbstractCompositionComponent} for the latter's
 * usual boilerplate). Either way, the component exposes its {@link CompositionComponentContext} via
 * {@link #context()}; everything else here is a default method built on top of it.
 *
 * <p>A component's template path is declared with {@link Composition}; a class's props are declared with
 * {@link Prop} on its fields; a record's props are its record components, one for one, needing no
 * declaration at all.
 */
public interface CompositionComponent {

    String DEFAULT_SLOT = "";

    /** This component occurrence's context — attributes, slot names, locale, messages, variables. */
    CompositionComponentContext context();

    default Map<String, Object> restAttributes() {
        return context().attributes().rest();
    }

    default boolean hasSlot() {
        return hasSlot(DEFAULT_SLOT);
    }

    default boolean hasSlot(String slotName) {
        return context().slotNames().contains(slotName);
    }
}

File: src/main/kotlin/blynx/thymeleaf/compositiondialect/CompositionComponent.kt
package blynx.thymeleaf.compositiondialect

import java.util.Locale

class TrackingAttributes(private val raw: Map<String, Any?>) {
    private val accessed = mutableSetOf<String>()

    operator fun get(name: String): Any? {
        accessed.add(name)
        return raw[name]
    }

    fun containsKey(name: String): Boolean {
        accessed.add(name)
        return raw.containsKey(name)
    }

    fun rest(): Map<String, Any?> = raw.filterKeys { it !in accessed }
}

class CompositionComponentContext(
    val attributes: TrackingAttributes,
    val slotNames: Set<String>,
    val locale: Locale,
    private val messageResolver: (String, Array<out Any?>) -> String?,
    private val variableReader: (String) -> Any?,
    private val variableWriter: (String, Any?) -> Unit,
) {
    fun message(code: String, vararg params: Any?): String = messageResolver(code, params) ?: code
    fun variable(name: String): Any? = variableReader(name)
    fun setVariable(name: String, value: Any?) = variableWriter(name, value)
}

open class CompositionComponent(
    private val componentContext: CompositionComponentContext,
) {
    companion object {
        const val DEFAULT_SLOT: String = ""
        const val path: String = ""
    }

    val restAttributes: Map<String, Any?>
        get() = componentContext.attributes.rest()

    @JvmOverloads
    fun hasSlot(slotName: String = DEFAULT_SLOT): Boolean = componentContext.slotNames.contains(slotName)
}
